File: core/cortex.ts
// core/cortex.ts
import { AM_Organism } from "./organism"
import * as math from "mathjs"

// ================= Symbol Metrics =================
class SymbolPhaseMetrics {
  usages: Record<string, number> = {}
  tasks: Record<string, Set<string>> = {}
  transitions: Record<string, Set<string>> = {}
  reusable: Record<string, boolean> = {}
  generalizes: Record<string, boolean> = {}
  epoch = 0

  observe(sym: string, task: string, prev: string) {
    this.usages[sym] = (this.usages[sym] || 0) + 1

    this.tasks[sym] ||= new Set()
    this.tasks[sym].add(task)

    this.transitions[sym] ||= new Set()
    if (prev) this.transitions[sym].add(prev)

    if (this.usages[sym] > 10 && this.transitions[sym].size > 3)
      this.generalizes[sym] = true

    if (this.usages[sym] > 7 && this.tasks[sym].size > 2)
      this.reusable[sym] = true
  }

  snapshot() {
    return Object.keys(this.usages).map(s => ({
      symbol: s,
      usage: this.usages[s],
      reusable: !!this.reusable[s],
      generalizes: !!this.generalizes[s],
      tasks: [...(this.tasks[s] || [])],
      transitions: [...(this.transitions[s] || [])]
    }))
  }
}

// ================= Symbol Graph =================
class SymbolNode {
  id: string
  vector: number[]
  parent: string | null = null
  children = new Set<string>()
  constructor(id: string, v: number[]) { this.id = id; this.vector = v }
}

class HierarchicalSymbolGraph {
  nodes: Record<string, SymbolNode> = {}
  programs: Record<string, string[]> = {}

  addSymbol(id: string, v: number[]) {
    if (!this.nodes[id]) this.nodes[id] = new SymbolNode(id, v)
  }

  addEdge(p: string, c: string) {
    if (this.nodes[p] && this.nodes[c]) {
      this.nodes[p].children.add(c)
      this.nodes[c].parent = p
    }
  }

  compose(meta: string, seq: string[]) {
    this.programs[meta] = seq
  }

  hierarchy() {
    return Object.fromEntries(
      Object.entries(this.nodes).map(([k,n]) => [
        k,
        { parent: n.parent, children: [...n.children] }
      ])
    )
  }
}

// ================= Counterfactual Planner =================
class CounterfactualPlanner {
  organism: AM_Organism
  graph: HierarchicalSymbolGraph

  constructor(org: AM_Organism, graph: HierarchicalSymbolGraph) {
    this.organism = org
    this.graph = graph
  }

  private rollout(start: number[], meta: string, depth = 6) {
    let z = start.slice()
    let score = 0

    const seq = this.graph.programs[meta] || []

    for (let i = 0; i < depth; i++) {
      const a = Array.from({ length: this.organism.actionDim }, () => Math.random() * 2 - 1)
      z = this.organism.predict(z, a)

      const util = math.sum(z) as unknown as number
      const stab = math.std(z) as unknown as number

      score += util - 0.2 * stab
    }

    return score
  }

  plan(z: number[]) {
    const metas = Object.keys(this.graph.programs)
    if (!metas.length) return null

    return metas
      .map(m => ({ meta: m, score: this.rollout(z, m) }))
      .sort((a,b) => b.score - a.score)[0]
  }
}

// ================= Cortex =================
export class AM_Cortex {
  organism: AM_Organism
  phase = new SymbolPhaseMetrics()
  graph = new HierarchicalSymbolGraph()
  planner: CounterfactualPlanner

  constructor(org: AM_Organism) {
    this.organism = org
    this.planner = new CounterfactualPlanner(org, this.graph)
  }

  observe(sym: string, task: string, prev: string, z: number[]) {
    this.phase.observe(sym, task, prev)
    this.graph.addSymbol(sym, z)
    if (prev) this.graph.addEdge(prev, sym)
  }

  compose(symbols: string[]) {
    const id = "M_" + Math.random().toString(36).slice(2,7)
    this.graph.compose(id, symbols)
    return id
  }

  plan(z: number[]) {
    return this.planner.plan(z)
  }

  metrics() { return this.phase.snapshot() }
  hierarchy() { return this.graph.hierarchy() }
}

File: python/cortex.py
import torch
import numpy as np
import math as pymath
import random

class SymbolPhaseMetrics:
    def __init__(self):
        self.usage = {}
        self.transitions = {}
        self.tasks = {}
        self.birth = {}
        self.death = {}
        self.reusable = {}
        self.general = {}
        self.epoch = 0

    def observe(self, sym, task, nxt):
        self.usage[sym] = self.usage.get(sym, 0) + 1
        self.tasks.setdefault(sym, set()).add(task)
        self.transitions.setdefault(sym, set()).add(nxt)
        self.birth.setdefault(sym, self.epoch)
        if self.usage[sym] > 15 and len(self.transitions[sym]) > 3:
            self.general[sym] = True
        if self.usage[sym] > 8 and len(self.tasks[sym]) > 2:
            self.reusable[sym] = True

    def collapse(self, sym):
        self.death[sym] = self.epoch

    def snapshot(self):
        return [{
            "symbol": s,
            "usage": self.usage[s],
            "general": self.general.get(s, False),
            "reusable": self.reusable.get(s, False),
            "tasks": list(self.tasks.get(s, [])),
            "born": self.birth.get(s),
            "dead": self.death.get(s)
        } for s in self.usage]

class HierarchicalSymbolGraph:
    def __init__(self):
        self.nodes = {}
        self.parents = {}
        self.programs = {}

    def add_symbol(self, sid, vec):
        self.nodes[sid] = np.array(vec)

    def add_edge(self, parent, child):
        self.parents[child] = parent

    def compose(self, mid, seq):
        self.programs[mid] = seq

    def execute(self, mid, z):
        for sid in self.programs.get(mid, []):
            z = self.nodes[sid]
        return z

class CounterfactualPlanner:
    def __init__(self, kernel):
        self.kernel = kernel
        self.rollout_width = 6
        self.rollout_depth = 6

    def _score_latent(self, z):
        entropy = -float(torch.sum(z * torch.log(z + 1e-8)))
        variance = float(torch.var(z))
        compression = -float(torch.mean(torch.abs(z)))
        return 0.6 * entropy + 0.3 * variance + 0.1 * compression

    def imagine(self, start_latent):
        best = None
        best_score = -1e9
        z0 = torch.tensor(start_latent, dtype=torch.float32)
        for _ in range(self.rollout_width):
            z = z0.clone()
            score = 0
            plan = []
            for _ in range(self.rollout_depth):
                a = torch.rand(self.kernel.action_dim) * 2 - 1
                z_next = self.kernel.wm(z.unsqueeze(0), a.unsqueeze(0)).squeeze(0)
                score += self._score_latent(z_next)
                plan.append(a.numpy().tolist())
                z = z_next.detach()
            if score > best_score:
                best_score = score
                best = plan
        return best

class AM_Cortex:
    def __init__(self, kernel, llm=None):
        self.kernel = kernel
        self.phase = SymbolPhaseMetrics()
        self.graph = HierarchicalSymbolGraph()
        self.planner = CounterfactualPlanner(kernel)
        self.llm = llm

    def observe(self, sym, task, nxt, vec):
        self.phase.observe(sym, task, nxt)
        self.graph.add_symbol(sym, vec)

    def plan(self, latent):
        return self.planner.imagine(latent)

    def metrics(self):
        return self.phase.snapshot()

    def compose_meta(self, symbols):
        meta_id = "M" + ''.join(random.choices("abcdefghijklmnopqrstuvwxyz0123456789", k=6))
        self.graph.compose(meta_id, symbols)
        return meta_id

    def hierarchy(self):
        return {k: {"parent": self.graph.parents.get(k), "vec": self.graph.nodes[k].tolist()} for k in self.graph.nodes}
